New vector graph implementation

szeli1 opened the pull request:
This pull request adds a new widget to LMMS called VectorGraph. It also includes an effect called "FFTFilter".

There was one vector graph implementation before this in pr #4367.

The widget is currently unfinished and the effect's functionality is not implemented.

The widget currently supports automation and a feature called "effector garph" is mostly implemented (not tested) enabling 2 or more graphs to interact with each other.
To automate a point on the graph, select the point and ctrl + drag to an automation track. Only 4 values called "point attributes" (there are more than 4 point attributes) are automatable, the user can decide what is automated by using the buttons shown in editing mode (when a point is clicked). A context menu is planned to enable controller support and hints. To see the automated values change, update the graph by pressing the right mouse button somewhere in the widget.
The graph supports different types of lines such as "sine", "peak", "steps", "random". Each line type can be modified using the (float) point attributes (curve, value A, value B).
In editing mode the graph's effector attributes are controllable as well. These effects are applied to an other graph which is configured to be effected by this graph.
There are multiple settings of a graph which are not controlled by the graph widget. These settings can be changed in code with the functions provided by the `VectorGraphModel` and the `VectorGraphDataArray` class. These settings include color and settings that apply to all the points in a `VectorGraphDataArray`.
`VectorGraphDataArray` contains all the point data, colors, settings, and helper vectors. The helper vectors are used to calculate the output line values in an optimized way (currently disabled). The "`getValues`" function calculates the output line values, by collecting all the changed points and recalculating their lines.

![Finished1](https://github.com/LMMS/lmms/assets/143485814/d065d189-1b85-4b3c-942b-b02d05c90ccf)
![Finished2](https://github.com/LMMS/lmms/assets/143485814/a3d98a2e-eec9-4a2e-ba2a-c2d552a638de)
![Finished3](https://github.com/LMMS/lmms/assets/143485814/262e66e7-f939-467e-b4aa-43669d5b4a96)
![Finished5](https://github.com/LMMS/lmms/assets/143485814/fa5b5327-5eb0-46f9-93ec-78d930e686ae)

TODO list:

- [ ] 	rewrite comments
- [ ] 	rename functions and values
- [ ] 	test all the features / settings
- [ ] 	do TODO-s

- [ ] 	automation:
- [ ] 	add automation support:
- [ ] 		save `FloatModels`
- [ ] 		safe deletion (currently not safe!!)

- [ ] 	event when a dataArray's size gets to 0
- [ ] 	ability to scale displayed coordinates in PointGraphView (not 0 - 100)		ignore?
- [ ]   scaled graph option in `VectorGraphView` which scales the last used output to be displayed in the widget (does not call `getValues`).
- [ ] 	call `dataChanged()` in `setDataArray`
- [ ] 	journalling in `PointGraphModel`
- [ ] 	`setDataArray` keep attributes option, `formatArray` option which runs `formatArray`
- [ ] 	context menu in gui (clear automation, connect to controller)
- [ ] 	display hints (full text) while editing
- [ ] 	ability to edit multiple graphs using `m_isLastSelectedArray`
- [ ] 	handle effector arrays when deleting `VectorGraphDataArray`
- [ ] 	update `formatArray`
- [ ] 	licensing email
- [ ] 	new button in gui which switches between the currently selected graphs
- [ ] 	dragging

- [ ]   filter implementation

![Finished4](https://github.com/LMMS/lmms/assets/143485814/13a1b203-dbed-4aae-97e8-c4eb777d25f4)

Keep in mind that this widget and the filter is **not finished**, so there could be harmful bugs/errors.
To access the new widget, search for an effect called "FFTFilter", open it and click the vector graph widget that should be in the middle (at first it does not show any gui).
Feel free to make suggestions and criticize my coding style / decisions.

Veratil reviewed (commented):
Haven't gone through most of `widgets/VectorGraph.cpp` but wanted to get these comments in now.

Veratil commented on src/gui/widgets/VectorGraph.cpp:
```diff
@@ -0,0 +1,4043 @@
+	return (inputValueIn / 2.0f + 0.5f) * displayLengthIn;
+}
+
+float VectorGraphView::getDistance(int xAIn, int yAIn, int xBIn, int yBIn)
+{
+	return std::sqrt(static_cast<float>((xAIn - xBIn) * (xAIn - xBIn) + (yAIn - yBIn) * (yAIn - yBIn)));
+}
+float VectorGraphView::getDistance(float xAIn, float yAIn, float xBIn, float yBIn)
+{
+	return std::sqrt((xAIn - xBIn) * (xAIn - xBIn) + (yAIn - yBIn) * (yAIn - yBIn));
+}
```
Or even a template maybe

szeli1 commented on src/gui/widgets/VectorGraph.cpp:
```diff
@@ -0,0 +1,4043 @@
+}
+VectorGraphView::~VectorGraphView()
+{
+	qDebug("VectorGraphView dstc");
```
I left those in for reviewers to better see what is happening. The whole getValues() (in VectorGraphDataArray) sample calculation takes up about 500 lines of code and could be difficult to understand. I'm going to remove those.

szeli1 commented on plugins/WaveShaper/WaveShaperControls.cpp:
```diff
@@ -44,26 +46,39 @@ WaveShaperControls::WaveShaperControls( WaveShaperEffect * _eff ) :
+	m_vectorGraphModel.getDataArray(arrayLocationB)->setIsSaveable(true);
+
+	unsigned int arrayLocation = m_vectorGraphModel.addArray();
+	//m_vectorGraphModel.getDataArray(arrayLocation)->setIsFixedSize(true);
+	//m_vectorGraphModel.getDataArray(arrayLocation)->setIsFixedX(true);
+	//m_vectorGraphModel.getDataArray(arrayLocation)->setIsFixedY(true);
+	//m_vectorGraphModel.getDataArray(arrayLocation)->setIsFixedEndPoints(true);
```
This is mostly an example implementation for VectorGraph, I left those in so if others want to implement it, they can see those options exist.

Veratil replied:
It's not obvious this is what that means. Make a block comment and describe that instead. It looks like this is leftover test code.

szeli1 replied:
This will be fixed then.

szeli1 commented on plugins/WaveShaper/WaveShaper.cpp:
```diff
@@ -82,7 +85,9 @@ bool WaveShaperEffect::processAudioBuffer( sampleFrame * _buf,
-	const float * samples = m_wsControls.m_wavegraphModel.samples();
+	// getting the current graph samples (size: 200)
+	// getting the graph samples this often can cause lagg with bigger sample count (if automated or effected)
+	std::vector<float> graphSamples = m_wsControls.getGraphSamples();
```
I will rework that. (and It is not because of that. It is because if the graph is automated then on every update a large amount of samples have to be recalculated)

szeli1 replied:
Updated this.

szeli1 commented on data/themes/default/style.css:
```diff
@@ -51,6 +51,17 @@ lmms--gui--Knob {
 	qproperty-arcInactiveColor: rgba(120, 120, 120, 70);
 }
 
+lmms--gui--VectorGraphView {
```
Thanks for noticing.

szeli1 commented on plugins/WaveShaper/WaveShaperControls.cpp:
```diff
@@ -72,14 +87,8 @@ void WaveShaperControls::loadSettings( const QDomElement & _this )
-	delete[] dst;
-
+	// loading VectorGraph
+	m_vectorGraphModel.loadSettings(_this, "VectorGraph"); 
```
It can not load old data currently I will change that.

szeli1 replied:
Added support for loading in old graph data.

szeli1 commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,729 @@
+	void setLineColor(QColor colorIn, unsigned int dataArrayLocationIn);
+	void setActiveColor(QColor colorIn, unsigned int dataArrayLocationIn);
+	void setFillColor(QColor colorIn, unsigned int dataArrayLocationIn);
+	void setAutomatedColor(QColor colorIn, unsigned int dataArrayLocationIn);
```
It is easier to understand what comes in as input in a function and without using In, variable names sound too general. It would be hard to replace "In" in all (roughly) 100 functions. I can do it if you want, but I would rather not replace those in `VectorGraph`.

Veratil replied:
Input to a function is defined by the arguments, you don't need to say it. The only case where it would make sense to use it is when there are arguments which are `out`.

I would rather we remove `In` everywhere.

szeli1 replied:
Resolved.

szeli1 commented on src/gui/widgets/VectorGraph.cpp:
```diff
@@ -0,0 +1,4043 @@
+{
+	if (m_isSelected == true)
+	{
+		qDebug("setSelectedData");
```
I can remove them if you want, but it will be harder to test.

szeli1 commented:
The core VectorGraphDataArray class needs to be reviewed.
The usage, construction and destruction of FloatModel-s need to be reviewed.
Function getSamples() (and related functinons) needs to be reviewed.
Spelling should be checked.

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	// clamps down the values to 0 - 1, -1 - 1
+	// sorts array, removes duplicated positions, calls dataChanged() if callDataChanged
+	// clamp: should clamp, sort: should sort
+	void formatArray(std::vector<std::pair<float, float>>* dataArrayOut, bool shouldClamp, bool shouldRescale, bool shouldSort, bool callDataChanged);
```
After seeing a lot of `std::pair<float, float>` I'm starting to think we should probably add a `using PointF = std::pair<float, float>` for simplicity.

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	public:
+		inline VectorGraphPoint()
+		{
+			m_x = 0.0f;
+			m_y = 0.0f;
+			m_c = 0.0f;
+			m_valA = 0.0f;
+			m_valB = 0.0f;
+			m_type = 0;
+			m_automatedEffectedAttribLocations = 0;
+
+			m_effectOnlyPoints = false;
+			m_effectAdd = false;
+			m_effectSubtract = false;
+			m_effectMultiply = false;
+			m_effectDivide = false;
+			m_effectPower = false;
+			m_effectLog = false;
+			m_effectSine = false;
+			m_effectClampLower = false;
+			m_effectClampUpper = false;
+
+			m_bufferedAutomationValue = 0.0f;
+			m_automationModel = -1;
```
These should all be set as default value where they're declared and not here in the constructor. Same below.

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+			m_bufferedAutomationValue = 0.0f;
+			m_automationModel = -1;
+		}
+		inline VectorGraphPoint(float x, float y)
+		{
+			m_x = x;
+			m_y = y;
```
After moving the defaults out:
```suggestion
		inline VectorGraphPoint(float x, float y) : m_x(x), m_y(y) {}
```

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	};
+	// deletes the point's automation model
+	// if modelLocation == point location
+	void delAutomationModel(unsigned int modelLocation, bool callDataChanged);
```
```suggestion
	void deleteAutomationModel(unsigned int modelLocation, bool callDataChanged);
```

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	// deletes the point's automation model
+	// if modelLocation == point location
+	void delAutomationModel(unsigned int modelLocation, bool callDataChanged);
+	// swapping values, "slide" moves the values (between) once left or right
+	// handle m_isFixedEndPoints when using this
+	void swap(unsigned int pointLocationA, unsigned int pointLocationB, bool slide);
```
What does slide do exactly? (I haven't gotten to where it's defined yet, but this doesn't tell me what it does still)

szeli1 replied:
A better name would be shouldOffsetPoints, or something similar, this will be changed. It basically moves/shifts the points between the 2 location to keep the order (ordered by x coords).

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	// handle m_isFixedEndPoints when using this
+	void swap(unsigned int pointLocationA, unsigned int pointLocationB, bool slide);
+	// returns the curve value at a given x coord, does clamp
+	float processCurve(float yBefore, float yAfter, float curve, float xIn);
```
```suggestion
	float processCurve(float yBefore, float yAfter, float curve, float x);
```

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	float processAutomation(unsigned int pointLocation, float attribValue, unsigned int attribLocation);
+
+	// line effects / types, m_type is used for this
+	// valA: amp, valB: freq, fadeInStartLoc: from what xIn value should the line type fade out
```
```suggestion
	// valA: amp, valB: freq, fadeInStartLoc: from what x value should the line type fade out
```

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+
+	// line effects / types, m_type is used for this
+	// valA: amp, valB: freq, fadeInStartLoc: from what xIn value should the line type fade out
+	//float processLineTypeSine(float xIn, float valA, float valB, float fadeInStartLoc);
+	std::vector<float> processLineTypeArraySine(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		float valA, float valB, float fadeInStartLoc);
+	// curve: phase
+	//float processLineTypeSineB(float xIn, float valA, float valB, float curve, float fadeInStartLoc);
+	std::vector<float> processLineTypeArraySineB(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		float valA, float valB, float curve, float fadeInStartLoc);
+	// valA: amp, valB: x coord, curve: width
+	//float processLineTypePeak(float xIn, float valA, float valB, float curve, float fadeInStartLoc);
+	std::vector<float> processLineTypeArrayPeak(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		float valA, float valB, float curve, float fadeInStartLoc);
+	// y: calculate steps from, valA: y count, valB: curve
+	//float processLineTypeSteps(float xIn, float yIn, float valA, float valBIn, float fadeInStartLoc);
+	std::vector<float> processLineTypeArraySteps(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		std::vector<float>* yArray, float valA, float valB, float fadeInStartLoc);
+	// valA: amp, valB: random number count, curve: seed
+	//float processLineTypeRandom(float xIn, float valA, float valB, float curve, float fadeInStartLoc);
+	std::vector<float> processLineTypeArrayRandom(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
```
The comments don't match the declaration anymore

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	// updating
+	// adds the points that are
+	// effected by the effector's values changing
+	// ONLY WORKS IN SORTED ARRAYS
```
And what happens if you pass a non-sorted array?

szeli1 replied:
It will not work. But the `effector` array's `m_needsUpdating` is always sorted.

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	// .first = start, .second = line end location (effector dataArray)
+	//void getSamplesLocations(VectorGraphDataArray* effector, std::vector<std::pair<unsigned int, unsigned int>>* effectorDataOut);
+	void getSamplesUpdateLines(VectorGraphDataArray* effector, std::vector<float>* effectorSamples,
+		std::vector<float>* outputXLocations, unsigned int iIn, float stepSize);
```
What is `i` in this function?

szeli1 replied:
I is m_dataArray[i] (where the points are stored). Will be renamed.

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	bool m_isFixedX;
+	// can the values be changed
+	bool m_isFixedY;
+	// if true then it makes the last position coordinate 1, 1, the first point coordinate to -1 (ot 0), 0
```
I assume `or`?
```suggestion
	// if true then it makes the last position coordinate 1, 1, the first point coordinate to -1 (or 0), 0
```

Veratil commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	std::vector<VectorGraphPoint> m_dataArray;
+	
+
+	// baking
```
Cookies? :)

szeli1 replied:
Got the name from texture baking. I guess what you are trying to say is this name does not fit in...

Veratil replied:
Oh, no. I was just leaving some fun commentary for a laugh. It's a big PR and going through a ton of lines can get mind numbing.
If the phase name makes sense, then it's fine to keep.

szeli1 replied:
Ok no problem. I appreciate your commentary. This is a big pr and even as someone who understands it reviewing all the comments and functions was tiring. Thanks for reviewing this.

Veratil commented on plugins/WaveShaper/WaveShaper.cpp:
```diff
@@ -119,17 +124,17 @@ bool WaveShaperEffect::processAudioBuffer( sampleFrame * _buf,
 			{
-				s[i] = linearInterpolate( samples[ lookup - 1 ],
-						samples[ lookup ], frac )
+				s[i] = linearInterpolate(graphSamples->operator[](lookup - 1),
```
Is there a reason you can't just do `(*graphSamples)[lookup - 1]` ?

szeli1 replied:
The type is std::vector<float>*, why use casting and use `[]` when `[]` can simply be used?

Veratil replied:
There is no casting. The difference is just using normal array access method.
```cpp
graphSamples->operator[](lookup - 1)
(*graphSamples)[lookup -1]
```
This accomplishes the same thing. It's just odd seeing calling `->operator[]()` directly.

szeli1 replied:
So replace all `operator[]()`? It is your decision.

Veratil replied:
I think `(*graphSamples)[lookup -1]` reads better.

Veratil commented on plugins/WaveShaper/WaveShaperControls.h:
```diff
@@ -63,26 +65,31 @@ class WaveShaperControls : public EffectControls
+	void vectorGraphChanged();
+signals:
+	// connected to dataArrayChanged()
+	// shouldUseGetLastValuesIn: when the graph is redrawn
```
```suggestion
	// shouldUseGetLastValues: when the graph is redrawn
```

Veratil commented on src/gui/widgets/VectorGraph.cpp:
```diff
@@ -0,0 +1,4112 @@
+#include <QMutex> // locking when getSamples
+
+
+#include "VectorGraph.h"
```
This should be the first include

Veratil reviewed (changes requested):
Still haven't gone over the 4000 lines VectorGraph.cpp yet...

szeli1 commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+{
+Q_OBJECT
+public:
+	VectorGraphModel(unsigned int arrayMaxLength, Model* parentIn, bool defaultConstructed);
```
I would like to keep `parentIn`. `parent` sounds too general to me and since this class inherits Q_Object, where `parent` could be a variable `parentIn` looks less confusing.

Veratil replied:
`parent` is the standard name for Qt objects. Use that please.

szeli1 replied:
All right, thanks for the answer.

szeli1 commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	bool m_isEditingActive;
+	std::vector<QString> m_controlText;
+	std::vector<QString> m_controlLineEffectText;
+	std::vector<bool> m_controlIsFloat;
```
Will std::array be right? Or should I use c style arrays?

Veratil replied:
If you know the size beforehand `std::array` will work. Do not use C-style arrays.
If we don't, then don't worry about changing this (yet).

szeli1 replied:
It is now std::array. Will tr() work?

Veratil replied:
I don't know what you're asking about tr?

szeli1 replied:
For example `std::array<QString, 18> m_controlText` uses tr() for translation. Will memory be allocated properly when the translated text is larger?

Veratil replied:
Why are you changing `m_controlText`? I was specifically talking about `std::vector<bool> m_controlIsFloat`, not about anything else.

szeli1 replied:
I am sorry I have accidentally misunderstood you. I think `m_controlText` should be changed from vector since it is never changed after construction. I just don't know how.

szeli1 commented on include/VectorGraph.h:
```diff
@@ -0,0 +1,739 @@
+	Q_PROPERTY(int fontSize MEMBER m_fontSize)
+public:
+
+	VectorGraphView(QWidget * parent, int widthIn, int heightIn, unsigned int pointSize,
```
Cant use height because of height() function. Renamed to widgetHeight.

szeli1 commented:
Added new class and moved the input dialog and the context menu there.
This does not effect any reviewed code up to this point. (but the new class needs to be reviewed)

szeli1 commented:
Most of the heap allocations were removed.
In the current configuration:
The best case for PaintEvent is 0 heap allocations.
The worst case for PaintEvent is 1 heap allocation.
The best case for updating the samples (getSamples()) is 0 heap allocations.
A reasonable worst case for updating the samples is 4 heap allocations.

This is after removing 9 heap allocations.

Added new button in WaveShaper for simplifying the graph. This can improve performance.

Rossmaxx commented on src/gui/widgets/VectorGraphViewBase.cpp:
```diff
@@ -0,0 +1,204 @@
+	menu->addAction(embed::getIconPixmap("controller"),
+		tr("Connect to controller..."),
+		this, SLOT(contextMenuExecConnectionDialog()));
+	if(m_curAutomationModel != nullptr && m_curAutomationModel->controllerConnection() != nullptr)
```
The other codefactor checks seems to be false positives but the one below seems to be an actual fail.

Rossmaxx reviewed (commented):
The changes are too large. I have reviewed it like i could. Also, consider splitting the file up (or using Qt/STL mechanisms where you can)

messmerd commented:
The controls seemed really unintuitive to me and I couldn't figure them out until I read the instructions in this PR, but even then it was a struggle.

The first thing I did was click and drag in an empty spot of the WaveShaper graph, expecting it to create a new node and move it around (like in the Automation Editor), but it behaved very differently.

There's text at the bottom of the WaveShaper graph showing the options, but it's not very legible and it's awkward to use. It took me a few minutes to realize that despite each option being a simple box, they function like knobs and you have to click and drag up or down to adjust them. I think it would be better to have a proper toolbar and some knobs like other editing windows in LMMS. If the size of the WaveShaper window needs to be increased to accommodate it, that's fine.

I also don't understand how the white waveform works and why it is needed.

I did notice a bug when I went to modify a Y coordinate of a node - I clicked "Cancel" in the input box but it modified the node anyway, setting it to zero. However, I wasn't able to replicate it. When I tried again, there were were two consecutive input boxes that appeared and I'm not sure why or what data it was editing.

Overall I'm impressed with the functionality the new vector graph widget provides, and being able to automate the nodes is awesome. I just think it needs some work to improve the user experience.

Rossmaxx commented:
It just randomly crossed my mind. How about using some functions from the automation track instead of implementing everything yourself (i believe there might be atleast a 10% similarity between both)

szeli1 commented:
> I also don't understand how the white waveform works and why it is needed.

Different waveforms or graphs can interact with each other. In this case the white graph can change the green one.
Currently this can help in automation:
We want to automate multiple points on the green graph slightly differently, instead of making multiple automation clips, we can automate a point on the white graph and in the settings of the green graph, we can enable the option "add", which would add the white graph's values to the green graph.

But in other configurations, an (input) graph might be used to display a signal for example. In that case the output graph can react to the input graph, changing the output values accordingly.

Older example:
![Finished9](https://github.com/LMMS/lmms/assets/143485814/e64d04d9-a4bb-421f-9838-79f7715495cb)
Here the blue graph is added (between 1. and 2. point) to the red, and then multiplied by the red (after the 2. point)


> I think it would be better to have a proper toolbar and some knobs like other editing windows in LMMS

I might add a new window for editing graphs.

szeli1 commented:
> consider splitting the file up (or using Qt/STL mechanisms where you can

What do you mean by "splitting the file up" and "Qt/STL mechanisms"?

szeli1 commented:
> It just randomly crossed my mind. How about using some functions from the automation track instead of implementing everything yourself (i believe there might be atleast a 10% similarity between both)

Probably not possible. The 2 systems are really different currently. But I might copy some things from the Automation Clip.

Rossmaxx commented:
> splitting the file up

The file is too large, might as well split up the functionality to a base file and helper functions in another file. If it ain't possible, leave it as is. 

> Qt/STL mechanisms

There might be something already possible through Qt/STL functions which you seemed to have missed out on. I haven't looked too close on the logic but that's what i feel from a quick glance. Lemme look again.

szeli1 commented:
> There might be something already possible through Qt/STL functions which you seemed to have missed out on. I haven't looked too close on the logic but that's what i feel from a quick glance. Lemme look again.

I noticed I've reimplemented some kind of sorted map, but Qt does not provide the necessary functions that I use for optimization. (such as finding the nearest x position of a point)

szeli1 commented:
I have added bezier curves, reviews are welcome especially for workflow and `VectorGraphModel`

bratpeki commented:
I'll be testing this. We need the waveshaper updated ASAP.

bratpeki commented:
Before anything, I would like to ask if we could make the waveshaper window bigger? At least 150% of the current size. As of now, it is ridiculously tiny.

Secondly, before a PR like this is made, it needs a cleaner UI.

![](https://private-user-images.githubusercontent.com/143485814/323723394-22dab93d-cdbe-4460-b5d8-64ca47f12555.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3MDc5NDEsIm5iZiI6MTcyNDcwNzY0MSwicGF0aCI6Ii8xNDM0ODU4MTQvMzIzNzIzMzk0LTIyZGFiOTNkLWNkYmUtNDQ2MC1iNWQ4LTY0Y2E0N2YxMjU1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNlQyMTI3MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YWM5YWU5NmVhYTIyNzUwNTM4ZTliZTEwZWEzM2U0Yzk5NGU2NGIxOWNjOTJjZWMzNTA2MDgwMTM5ZGQ5MThmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.uJOrfF3_SiL16jjrlG7M0IeGTFSm_pS4Yu6H9aLS95g)

What is this white line at the bottom? Why is it there?

![screencap-2024-08-26-23-36-31](https://github.com/user-attachments/assets/c6c71d51-33ac-4853-b05c-a9c615468caa)

Why is the text not at least a bit padded. Also, the text in the middle is pretty difficult to read.

---

So, for starters, I would argue this needs a bigger UI, and a bit of a color rework.

Rossmaxx commented on include/VectorGraphView.h:
```diff
@@ -0,0 +1,239 @@
+	QColor m_vectorGraphSecondaryActiveColor;
+	QColor m_vectorGraphSecondaryFillColor;
+
+	friend class lmms::gui::VectorGraphCotnrolDialog;
```
```suggestion
	friend class lmms::gui::VectorGraphControlDialog;
```

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+#include "MainWindow.h" // getting main window for control dialog
+#include "ProjectJournal.h"
+
+//#define VECTORGRAPH_DEBUG_USER_INTERACTION
```
Or wrap this in `#ifdef LMMS_DEBUG` or some similar.

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+void VectorGraphModel::saveSettings(QDomDocument& doc, QDomElement& element, const QString& name)
+{
+#ifdef VECTORGRAPH_DEBUG_PAINT_EVENT
+	qDebug("saveSettings start");
```
Consider using std::cerr instead in core. Change other occurences too.

szeli1 replied:
I changed it to printf().

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+	// generating 2 seeds and blending in between them
+	for (unsigned int i = 0; i < randomValuesSize / 2; i++)
+	{
+		m_universalSampleBuffer[i] = std::fmod((static_cast<float>(rand()) / 10000.0f), 2.0f) - 1.0f;
```
For rand, there's a fast_rand() in lmms_math.h. use that.

szeli1 replied:
I must use random functions with seeds.

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+	float blend = 10.0f + randomSeed * 10.0f;
+	int randomSeedB = static_cast<int>(blend);
+	blend = blend - randomSeedB;
+
```
No need to seed if using fast_rand.

szeli1 replied:
Seeds are used to create the desired behavior.

Rossmaxx reviewed (commented):
Half a review done. I'll try reviewing the rest soon. Fix the codefactor whitespace checks.

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+	dataChanged();
+}
+
+void VectorGraphDataArray::formatArray(std::vector<PointF>* dataArrayOut, bool shouldClamp, bool shouldRescale, bool shouldSort, bool callDataChanged)
```
you can break this function down into sortArray, RescaleArray, clampArray, and call these helpers from within this function, will get rid of codefactor warning.

szeli1 replied:
The order of these operations can not be changed and the `dataChanged()` will have to be implemented and future users of these functions would have to understand when to call `dataChanged()`.

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+unsigned int VectorGraphDataArray::setX(unsigned int pointLocation, float newX)
+{
+	int location = pointLocation;
+	if (m_isFixedX == false && newX <= 1.0f)
```
would make sense to invert this too.
```suggestion
	if (m_isFixedX == true || newX > 1.0f)
```

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+			int targetLocation = location;
+			// bool dataChangedVal = false;
+			// if getNearestLocation returned a value
+			if (location >= 0)
```
another place where inversion might help

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+			}
+			else
+			{
+				location = pointLocation;
```
if you decide on inverting the conditions, you can return pointLocation directly in these 2 areas.

szeli1 replied:
That condition is complex but readable, but if inverted it will be hardly understandable.

Rossmaxx commented on src/core/VectorGraphModel.cpp:
```diff
@@ -0,0 +1,2411 @@
+	float output = attribValue;
+
+	// none
+	if (getEffect(pointLocation, effectSlot) == 0) { return output; }
```
switch case exists.

Rossmaxx reviewed (commented):
More follow up reviews needed, here's the second

szeli1 commented:
> Secondly, before a PR like this is made, it needs a cleaner UI.
> Why is the text not at least a bit padded. Also, the text in the middle is pretty difficult to read.

Few pictures show outdated GUI.

> What is this white line at the bottom? Why is it there?

In the screenshot in your comment the white line acts as a clamp for the green line.
I will rewrite the PR description because nobody seems to understand this "white line" feature.

bratpeki commented:
I'm testing it now.

---

Why is this line so broken?

![screencap-2024-09-03-17-26-39](https://github.com/user-attachments/assets/2b492b5c-45c6-4f73-9764-8ce9b1b8e0b8)

Zoomed in:

![image](https://github.com/user-attachments/assets/52bb6d6d-6308-40ba-b91c-a385ae23d390)

---

I'm sending a video of me trying the feature out, along with some timestamps and questions. Ignore the Discord notification.

https://github.com/user-attachments/assets/e1f438f0-9315-4bf6-864c-a479a9d373e6

0:02 : The way the buttons at the bottom pop up is very unexpected. The user wants to move the point without some options popping up and changing the curve from its original shape.

0:03 : When the button at the bottom activate, the background doesn't follow them. Example:

![screencap-2024-09-03-17-42-53](https://github.com/user-attachments/assets/cfebf5a2-713c-4e1c-b5e9-83ae6aca0708)

The buttons also can't be closed with `Esc`. The curve looks like it has tiny "saw teeth", instead of being completely straight.

0:11 : You said this PR is done, right? Why do the buttons stick so close to the left? Why are they lowercase? Why does the Simplify" button have different capitalization depending on the letter? Why is its button so small? It can't fit all those letters. I think "switch graph" doesn't explain much. What is that other graph? Why use it?

0:13 : So, clicking to create a new point removes the buttons from the bottom, and clicking to change the button position opens it again. If I just wanted to move a point, I don't think I'd want that to pop up. It seems like an "advanced functionality". I also feel like that should be closeable, so I can return to editing points. 

0:19 : The point placement is very nice. I don't know how, but I got one of the pins moving the wrong direction relative to the mouse! 🤣

0:24 : Also, while moving pins, the curve "twitches". 

0:48 : Also twitches for that white curve thing.

1:00 : This really can't be called a finished PR with this UI.

1:21 : I tried giving the point a value. I'm expecting that the x value is cemented, and you edit the Y value by double-clicking, much like the lollipops in the MIDI editor, or the automation editor point. However it opens two inputs? Are those for X and Y?

---

Not regarding the video, this doesn't look quite right, but I may be wrong with this and maybe it should really be like this:

![screencap-2024-09-03-17-46-11](https://github.com/user-attachments/assets/fe4bf3af-e9bc-4e81-bdd8-9b0da5faf060)
![screencap-2024-09-03-17-46-18](https://github.com/user-attachments/assets/8393e8cd-7c3e-4ee0-9f6a-1a3062aaa972)

Also, why is there just a random 0 in the top-left??? 😂

---

I'm waiting on responses, but this now seems like either a very complex plugin, or a plugin that still demands a fair bit of work. As for the vector graph implementation, that's a bit too technical for me, so I can't comment on it.

szeli1 commented:
> Why is this line so broken?

It is probably because 200 samples are mapped to the 206 sample display. This is an optimization the graph uses to save recomputing the sample points.

> 0:02 : The way the buttons at the bottom pop up is very unexpected. The user wants to move the point without some options popping up and changing the curve from its original shape.

How would you like the options to show up? The user can use the editing window if they don't want to see the options.

> 0:03 : When the button at the bottom activate, the background doesn't follow them. Example:

Yes. I might add an option to scale the background to the correct size when the options show up.

> The buttons also can't be closed with `Esc`.

I will change how the shortcuts work in a future PR. I think this is an unnecessary shortcut since the user can press the right mouse button to end the point selection.

> 0:11 : You said this PR is done, right? Why do the buttons stick so close to the left? Why are they lowercase? Why does the Simplify" button have different capitalization depending on the letter? Why is its button so small? It can't fit all those letters.

> I think "switch graph" doesn't explain much.

It changes which graph is edited.

> What is that other graph?

Explained in PR body and in other comments.

> Why use it?

Explained in PR body and in other comments.

> 
> 0:13 : So, clicking to create a new point removes the buttons from the bottom, and clicking to change the button position opens it again. If I just wanted to move a point, I don't think I'd want that to pop up. It seems like an "advanced functionality". I also feel like that should be closeable, so I can return to editing points.

How to open that window without covering the graph in some way while not hiding the options?
I think the scaling issue is not that big. I will make it so when the points are moved, the options menu will close.


> 0:19 : The point placement is very nice. I don't know how, but I got one of the pins moving the wrong direction relative to the mouse! 🤣

I will look into that.

> 0:24 : Also, while moving pins, the curve "twitches".
> 0:48 : Also twitches for that white curve thing.

This is again because of the sample difference. Editing will cause it to draw 206 samples while the plugin uses 200 samples, so when the plugin updates the displayed graph will change. I don't consider this a big issue.

> 1:00 : This really can't be called a finished PR with this UI.

I think the UI is 95% finished.

> 1:21 : I tried giving the point a value. I'm expecting that the x value is cemented, and you edit the Y value by double-clicking, much like the lollipops in the MIDI editor, or the automation editor point. However it opens two inputs? Are those for X and Y?

Yes, it makes sense that you think that. I could change that if you want. At least the input dialogs should display which value you are editing.

> Not regarding the video, this doesn't look quite right, but I may be wrong with this and maybe it should really be like this:

I don't know what are you trying to show. The curve is mirrored but it displays correctly.

> Also, why is there just a random 0 in the top-left??? 😂

Explained in PR body and in other comments.
It displays which graph has focus currently.

> I'm waiting on responses, but this now seems like either a very complex plugin

I would like to mention that I think you didn't really test the features this graph offers. Keep in mind that this graph isn't for this plugin only, it is like the existing `Graph` Your comment might be true because I didn't write a good enough description for it, but In my view calling it complex doesn't mean a lot when you don't know how it works. An inexperienced user might call LMMS complex and unnecessary while an experienced user knows how things work and why they work like that. Being complex isn't a drawback when the basic graph has a learning curve of 20 seconds. The user doesn't have to use the complex features, but if those features are missing, then all users will be limited.

An information button would get rid of the "complex" argument tho.

regulus79 commented:
The options which appear at the bottom of the vector graph--would it look better if they were in a right-click context menu or something? Currently to me the way they don't have a background and take up a sizable portion of the graph doesn't seem quite right, idk.

(disclaimer, I haven't actually tested the pr yet. I should really do that, sorry)

szeli1 commented on include/VectorGraphModel.h:
```diff
@@ -0,0 +1,563 @@
+	friend class lmms::gui::VectorGraphView;
+};
+
+class LMMS_EXPORT VectorGraphDataArray
```
I would rather not do this, it will make it harder to read

szeli1 commented on include/VectorGraphModel.h:
```diff
@@ -0,0 +1,563 @@
+class VectorGraphDataArray;
+class FloatModel;
+
+using PointF = std::pair<float, float>;
```
`VectorGraphView` is no longer included.

messmerd commented:
I haven't tested this recently, and the automated builds are currently broken due to the merge conflict, but reading the PR description, it doesn't appear that much changed in terms of functionality and appearance since I last tried it, other than the addition of the "vector graph settings" window.

By far the biggest problem with this PR in my opinion is still how strange, unintuitive, and bad-looking the UI is, which leads to poor UX. And I'm not the only one to point this out. It's not enough for a UI to provide the minimum necessary functionality - it also has to look presentable and be easy to use.

Here are a few of the issues
- The "buttons" inside the graph (the context menu?) do not look like clickable buttons - they are just plain text with a couple lines drawn around them to form a box.
- The "switch graph" button is accessed through the context menu that appears when you click on a node, even though it has nothing to do with the individual node. If the vector graph is to have the modulation curve (which is another conversation), it should have a UI element that makes sense for it and is always present, such a tab widget to switch between views or a toggle button in the corner with an appropriate icon.
- The new "vector graph settings" window looks very disorganized and contains typos. It's accessed through the context menu for the currently selected node, but appears to contain features that affect both the selected node and the graph as a whole (though I could be wrong). I think actions that apply to a node should not need separate windows unless absolutely necessary. The Automation Editor, for example, has a toolbar with several buttons that allow editing different parameters of nodes. Imagine if a user wanted to edit the tangent of an automation node but the Automation Editor required them to click the node, then click the "edit point" text that pops up at the bottom, then adjust a knob in a separate "automation graph settings" window rather than having an "edit tangents" tool that they can use. That would be a really bad experience for users.

As far as the functionality this PR offers, the UI prevented me from being able to test or understand a lot of it, and I'm not sure how the functionality compares to the vector graphs of other DAWs. If there's some niche or oddly-implemented functionality, it should probably be removed or reworked so that it doesn't stray far from what is tried-and-true. Remember that we can always add features in the future, but we cannot easily remove features after they've been released, so the features we do add need to be carefully designed and broadly useful.

The secondary curve (the white curve used for modulation) may be such a feature to reconsider. Maybe the UI is partly to blame for souring me on the idea, but it does significantly complicate the PR and I'm not sure that it's the best way to implement modulation. For example, maybe the ability to link two vector graphs together could to the same thing in a more modular and extensible way while also being more efficient for people who don't need the 2nd curve? And in the future, if LMMS gained a [modulation system like Bitwig](https://www.bitwig.com/learnings/an-introduction-to-modulators-45/) maybe it would allow them to do many of the same things in a better way?

Overall I think there's still a lot of work that needs to be done on this PR, and it makes me wonder whether that effort might be better spent simply finishing #4367 which appears to have a good UI and does what it needs to without being overwhelmed by extra features.

sakertooth commented:
Just wanted to drop my own take as well: The nearly 6000 lines of code added is giving me bad vibes right off the bat. I wonder if we can take the problem we are trying to solve here and break it down, then merge smaller PRs incrementally? If there is any other way to not do this all at once... then I ask that we please do so. It will help with reviewing the code properly and making sure we do this right.

szeli1 commented:
> that much changed in terms of functionality and appearance since I last tried it, other than the addition of the "vector graph settings" window

I forgot to mention that i reworked how the waveshaper plugin looks.

> By far the biggest problem with this PR in my opinion is still how strange, unintuitive, and bad-looking the UI is, which leads to poor UX. And I'm not the only one to point this out. It's not enough for a UI to provide the necessary functionality - it also has to look presentable and be easy to use.

Would it be better if the controls were similar to the automation editor?
How to make it persentable? What makes it "bad looking"? The controls are based on Piano roll, so I don't really understand what is the problem. What makes the automation editor better in terms of ui? It is easy to use for me...

> The buttons inside the graph (the context menu?) do not look like clickable buttons - they are just plain text with a couple lines drawn around them to form a box.

Okay, this is something concrete I can fix.

> The "switch graph" button is accessed through the context menu that appears when you click on a node, even though it has nothing to do with the individual node. If the vector graph is to have the modulation curve (which is another conversation), it should have a UI element that makes sense for it and is always present, such a tab widget to switch between views or a toggle button in the corner with an appropriate icon.

I could fix this by reintroducing the context menu. It will show different settings when a point is clicked or when it isn't. This will minimise the space it takes up and solves the "doesn't look like a button" problem.

> The new "vector graph settings" window looks very disorganized and contains typos.

What are the typos?

> but appears to contain features that affect both the selected node and the graph as a whole (though I could be wrong).

Yes this is wrong, it only contains point specific settings.

> The Automation Editor, for example, has a toolbar with several buttons that allow editing different parameters of nodes.

Are you requesting me to make a graph editor? How would you fix this? Each point has many different settings, so I think this control dialog window is useful. I prefer this approach since allows precision while users have to guess what the node attributes are with tools. An other option would be placing the 2 attribute knobs below the widget, so all basic edits are available without opening the settings dialog. An other improvement would be if the settings dialog didn't close when the graph is clicked. How would you feel about these changes?

> The secondary curve (the white curve used for modulation) may be such a feature to reconsider. Maybe the UI is partly to blame for souring me on the idea, but it does significantly complicate the PR and I'm not sure that it's the best way to implement modulation. For example, maybe the ability to link two vector graphs together could do the same thing in a more modular and extensible way while also being more efficient for people who don't need the 2nd curve? And in the future, if LMMS gained a modulation system like Bitwig maybe such a modulation system could allow them to do many of the same things in a better way?

It is how it's done in code, they are linked together in waveshaper. I will remove the modulation feature to make the PR smaller.

Sorry for typos

What do you think about these ideas to improve this pr? @messmerd

szeli1 commented:
> What is the advantage of this PR over

This PR does these additional things:
1. can display multiple graphs in a single widget
2. can be automated
3. supports modulation

> PR 4367 is a third the size of this PR

If modulation is removed, I can get rid of 2000 lines probably

> has a clean and coherent UI

You can only do so much with that UI. Automation is a key factor why the gui looks different.
My PR offers additional advanced functionality that the user Isn't forced to use or understand.

> easy-to-understand API

I believe the same can be said about this PR If I clean up some things.

> and even documentation

This can also be said about this PR, the API isn't documented, but documentation is available to the end user inside lmms.

> FL Studio's WaveShaper which is an established design that appears to be well-liked by its users, so I can be more confident that its set of features is well designed and we won't have issues supporting it into the future.

I believe all of the things that graph does is supported by this PR. The gui could be better it is true.

I would like a graph that supports automation. From a plugin developer's perspective, displaying multiple kinds of information in a single widget is also useful (like displaying FFT peaks in an equalizer). I would like to remove the modulation functionality from this PR.

If you have problems with the GUI or a better vision, please let me know. The gui is the weakest part of this PR and I'm willing to implement every suggestion.

Gregaras commented:
Maybe graph buttons could be outside the graph like this so it would not make the graph view move:
[buttons_new.webm](https://github.com/user-attachments/assets/3a1515fa-2d54-40fa-9b11-dcbf5002529b)
And the vector graph setting window could maybe look like this:
<img width="312" height="528" alt="image" src="https://github.com/user-attachments/assets/55131f47-6618-45bd-a80f-23c0758ed5f0" />
Perhaps it could be possible to make the "effect point" and "effect line" buttons change color or something to be able to see whether they are in on or off state.
Maybe the left mouse button could just move points around and not create them, right mouse button could create them instead and if you right-click on the point, a dialog could appear with delete button in it (like in [this waveshaper](https://github.com/LMMS/lmms/pull/4367))

Whenever I put the effect on a sinewave, it distorts it.
Also, it seems that if I delete the last point of the white line and click on the point of the green line, I can no longer create white line points.
